proposals/2023/03-16_scheduled-builds.md

```diff
@@ -0,0 +1,287 @@
+### YAML
+
+A `schedule` event type will be added under `ruleset` for the new `schedules` functionality.
+
+This will instruct Vela to execute the step if the `event` for a build is `schedule`.
+
+The `target` field under a `ruleset` can be used to control what schedule a step will run for.
+
+This will enable further customization when multiple schedules are setup for a repo.
+
+```yaml
+steps:
+  - name: unit-test
+    ruleset:
+      event: [ deployment, schedule ]
+      target: [ unit_test, integration_test, nightly ]
+    commands:
+      - echo "I run when a schedule with the name 'unit_test', 'integration_test' or 'nightly' is executed"
+
+  - name: integration-test
+    ruleset:
+      event: [ deployment, schedule ]
+      target: [ integration_test, nightly ]
+    commands:
+      - echo "I run when a schedule with the name 'integration_test' or 'nightly' is executed"
+      
+  - name: publish
+    ruleset:
+      event: [ deployment, schedule ]
+      target: [ nightly ]
+    commands:
+      - echo "I run when a schedule with the name 'nightly' is executed"
+      
+  - name: notification
+    ruleset:
+      event: [ deployment, schedule ]
+    commands:
+      - echo "I run when any schedule is executed"
+```
+
+> NOTE: This leverages the existing `target` field but we could add a new field i.e. `schedule_target`.
+>
+> However, we elected to use `target` to reduce the complexity and code required to deliver on this feature.
+>
+> Also, this would offer the ability to trigger a build via a `schedule` or a `deployment` for various use cases.
+
+## Implementation
+
+<!--
+This section is intended to explain how the solution will be implemented for the proposal.
+
+NOTE: If there are no current plans for implementation, please leave this section blank.
+-->
+
+Users will be able to manage schedules for a repo via the UI, API and CLI.
+
+Only users with `admin` access to a repo will be able to manage schedules for that repo.
+
+Vela will support having multiple schedules setup for a repo.
+
+By default, when a repo is enabled in Vela it will not create a `schedule`.
+
+When a `schedule` is created for a repo, a new row will be added to a table for it.
+
+By default, Vela will support a minimum of `1hr` as the cadence for the schedule.
+
+However, similar to build limits and timeouts, Vela will support configuring that minimum via environment variable.
+
+The build rate limit per repo will protect the system from a DDOS effect for frequent schedules.
+
+Each build created from a schedule will leverage the configured `branch` of a repo for the pipeline (`.vela.yml`).
+
+The server will create a thread on startup to scan the table containing all schedules and spawn builds accordingly.
+
+The thread will be setup to scan the table every half of the minimum supported cadence i.e. `1hr` / 2 = `30m`
+
+The server will use [jitter](https://en.wikipedia.org/wiki/Jitter) to minimize the amount of builds created at the same time for all repos.
```

Comment:
When there's an outage (i.e. upgrades, disasters, etc), how would you anticipate Scheduled Builds working? Does it skip any "schedules" it missed or would it attempt to catch up? If we play "catch up", that may spawn a lot of builds immediately following an outage. Alternatively, if we skip the "schedules" how much scope creep do we have to add an ondemand re-run of the pipeline without a PR?

Reply:
@jbrockopp How about an on-demand re-run? We've seen this request in the past and, depending on how we approach Scheduled Builds, this could be implemented very similarily. Do you think a "Run Now" option could fit into this proposal?

Reply:
Yeah, I do think there is value in providing such a feature and we're hoping to leverage `deployments` for this:

https://github.com/go-vela/community/blob/feature/proposal/scheduled_builds/proposals/2023/03-16_scheduled-builds.md#yaml